### ISSUE TITLE
x86_64 +softfloat crashes LLVM

### ISSUE THREAD

**Issue by ominitay** [ziglang/zig#10847]
## First bad commit
9bbd3ab

## Steps to reproduce
Attempt to build [Zigvale Barebones](https://github.com/ominitay/zigvale-barebones) or [Florence OS](https://github.com/FlorenceOS/Florence/).

## Expected Behaviour
The projects build successfully. 

## Actual Behaviour
The compiler crashes in release mode, or segfaults in debug mode. 

cc: @N00byEdge 
cc: @Vexu

**Comment by ominitay** [comment]
Found the first bad commit with a bisect: 9bbd3ab

**Comment by ifreund** [comment]
Could you provide a stack trace?

**Comment by ominitay** [comment]
Unfortunately not. The only output I get is this:
```
Segmentation fault at address 0x0
???:?:?: 0x7fc10cca2a80 in ??? (???)
???:?:?: 0x9058def in ??? (???)
kernel...The following command terminated unexpectedly:
/home/renami/zig/build/zig build-exe /datadrive/Devel/zigvale-barebones/src/kernel.zig --cache-dir /datadrive/Devel/zigvale-barebones/zig-cache --global-cache-dir /home/renami/.cache/zig --name kernel -mcmodel kernel -target x86_64-freestanding-none -mcpu=x86_64-mmx+soft_float-sse-sse2 --script /datadrive/Devel/zigvale-barebones/src/linker.ld --pkg-begin zigvale /datadrive/Devel/zigvale-barebones/.zigmod/deps/git/github.com/Ominitay/zigvale/zigvale.zig --pkg-end --enable-cache
error: the following build command failed with exit code 6:
/datadrive/Devel/zigvale-barebones/zig-cache/o/51338842e1db1034ac75a325c125084a/build /home/renami/zig/build/zig /datadrive/Devel/zigvale-barebones /datadrive/Devel/zigvale-barebones/zig-cache /home/renami/.cache/zig
```

**Comment by ominitay** [comment]
Will do so now :)

**Comment by andrewrk** [comment]
The bug report template is there for a reason... please edit your issue to include all these fields: https://github.com/ziglang/zig/issues/new?assignees=&labels=bug&template=bug.yml

**Comment by andrewrk** [comment]
I see, ok I can try to follow the instructions on these projects' readme to build from source.

**Comment by davidgm94** [comment]
I get this stack trace while compiling my project (also a freestanding one). I will try to figure out what's going on but can't be sure since the crash is inside LLVM.

`Thread 1 "zig" received signal SIGSEGV, Segmentation fault.
0x00007ffff53bc080 in ?? () from /lib/x86_64-linux-gnu/libLLVM-13.so.1
(gdb) bt
#0  0x00007ffff53bc080 in ?? () from /lib/x86_64-linux-gnu/libLLVM-13.so.1
#1  0x00007ffff339caba in ?? () from /lib/x86_64-linux-gnu/libLLVM-13.so.1
#2  0x00007ffff3399996 in ?? () from /lib/x86_64-linux-gnu/libLLVM-13.so.1
#3  0x00007ffff32b596e in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () from /lib/x86_64-linux-gnu/libLLVM-13.so.1
#4  0x00007ffff30a698d in llvm::FPPassManager::runOnFunction(llvm::Function&) () from /lib/x86_64-linux-gnu/libLLVM-13.so.1
#5  0x00007ffff30ac333 in llvm::FPPassManager::runOnModule(llvm::Module&) () from /lib/x86_64-linux-gnu/libLLVM-13.so.1
#6  0x00007ffff30a6fdf in llvm::legacy::PassManagerImpl::run(llvm::Module&) () from /lib/x86_64-linux-gnu/libLLVM-13.so.1
#7  0x000055555716ced6 in ZigLLVMTargetMachineEmitToFile (targ_machine_ref=0x555559f125c0, module_ref=0x555559f0e3a0, error_message=0x7fffffff3030, is_debug=false, is_small=false, 
    time_report=false, tsan=false, lto=false, asm_filename=0x0, bin_filename=0x555559f0e220 "/home/david/.cache/zig/tmp/353b596f41897fac/compiler_rt.o", llvm_ir_filename=0x0, 
    bitcode_filename=0x0) at /home/david/git/zig/src/zig_llvm.cpp:389
#8  0x00005555570277c9 in zig_llvm_emit_output (g=0x555559f063e0) at /home/david/git/zig/src/stage1/codegen.cpp:9261
#9  0x000055555702c908 in codegen_build_object (g=0x555559f063e0) at /home/david/git/zig/src/stage1/codegen.cpp:10449
#10 0x0000555556ff9f82 in zig_stage1_build_object (stage1=0x555559f063e0) at /home/david/git/zig/src/stage1/stage1.cpp:132
#11 0x000055555652f695 in Module.build_object (mod=0x555559f063e0) at /home/david/git/zig/src/stage1.zig:149
#12 0x00005555564ffbcf in Compilation.updateStage1Module (comp=0x555559f02fe8, main_progress_node=0x7fffffff41d8) at /home/david/git/zig/src/Compilation.zig:5078
#13 0x00005555562cf85e in Compilation.processOneJob (comp=0x555559f02fe8, job=..., main_progress_node=0x7fffffff41d8) at /home/david/git/zig/src/Compilation.zig:3142
#14 0x00005555562c0ce3 in Compilation.performAllTheWork (self=0x555559f02fe8) at /home/david/git/zig/src/Compilation.zig:2676
#15 0x00005555562ba055 in Compilation.update (comp=0x555559f02fe8) at /home/david/git/zig/src/Compilation.zig:2082
#16 0x00005555564fd865 in Compilation.buildOutputFromZig (comp=0x555559effd28, src_basename=..., output_mode=Lib, out=0x555559f00070, misc_task_tag=compiler_rt)
    at /home/david/git/zig/src/Compilation.zig:4909
#17 0x00005555562cf21f in Compilation.processOneJob (comp=0x555559effd28, job=..., main_progress_node=0x7fffffff6298) at /home/david/git/zig/src/Compilation.zig:3063
#18 0x00005555562c0ce3 in Compilation.performAllTheWork (self=0x555559effd28) at /home/david/git/zig/src/Compilation.zig:2676
#19 0x00005555562ba055 in Compilation.update (comp=0x555559effd28) at /home/david/git/zig/src/Compilation.zig:2082
#20 0x00005555562408b0 in main.updateModule (gpa=..., comp=0x555559effd28, hook=...) at /home/david/git/zig/src/main.zig:2964
#21 0x000055555615a9e0 in main.buildOutputType (gpa=..., arena=..., all_args=..., arg_mode=...) at /home/david/git/zig/src/main.zig:2651
#22 0x0000555556124528 in main.mainArgs (gpa=..., arena=..., args=...) at /home/david/git/zig/src/main.zig:210
#23 0x0000555556123f3a in main (argc=19, argv=0x7fffffffe258) at /home/david/git/zig/src/stage1.zig:48`

**Comment by davidgm94** [comment]
I just tried with the last master. Unfortunately, the issue is still there. Do you want me to provide a new stack trace?

**Comment by ominitay** [comment]
Issue persists for me with `0.10.0-dev.812+5af9d0c60`.

**Comment by andrewrk** [comment]
The stack trace you provided tells us the issue is likely an LLVM bug. The next step towards solving the problem is to come up with a reduction; a way to trigger the crash with as simple a snippet of code as possible. Then we can work towards an LLVM IR reduction from there.

Once we are in LLVM IR territory, we can use the `llvm-reduce` tool.

**Comment by ominitay** [comment]
> I don't see a confirmation in either of the above two comments that specifically a CI build was used.

Just to confirm this, I've just downloaded the tarball from the [latest passing CI run](https://ci.ziglang.org/ziglang/zig/2552/1/3), and the issue is still there.

**Comment by ominitay** [comment]
> If the problem really is in compiler-rt then it should be as simple as an empty zig file.

@andrewrk The crash still happens with an empty Zig file.

**Comment by ominitay** [comment]
The minimal reproduction steps are then:
```
$ touch empty.zig
$ zig build-exe empty.zig -target x86_64-freestanding-none -mcpu=x86_64-mmx+soft_float-sse-sse2
segmentation fault (core dumped)  zig build-exe empty.zig -target x86_64-freestanding-none
```

**Comment by davidgm94** [comment]
On MacOS ARM, you don't get the same error as before (related to soft-float) with the example you just gave. You also have to specify -target if you are cross-compiling from another CPU architecture.

```david@Davids-MacBook-Air ~/git (mac) [SIGSEGV]> lldb -- zig-macos-aarch64-0.10.0-dev.865+8841a71aa/zig build-exe empty.zig -target x86_64-freestanding-none -mcpu=x86_64+soft_float
(lldb) target create "zig-macos-aarch64-0.10.0-dev.865+8841a71aa/zig"
Current executable set to '/Users/david/git/zig-macos-aarch64-0.10.0-dev.865+8841a71aa/zig' (arm64).
(lldb) settings set -- target.run-args  "build-exe" "empty.zig" "-target" "x86_64-freestanding-none" "-mcpu=x86_64+soft_float"
(lldb) r
Process 7593 launched: '/Users/david/git/zig-macos-aarch64-0.10.0-dev.865+8841a71aa/zig' (arm64)
Process 7593 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00000001027f466c zig`llvm::X86RegisterInfo::shouldRewriteCopySrc(llvm::TargetRegisterClass const*, unsigned int, llvm::TargetRegisterClass const*, unsigned int) const
zig`llvm::X86RegisterInfo::shouldRewriteCopySrc:
->  0x1027f466c <+0>:  ldr    x8, [x1]
    0x1027f4670 <+4>:  ldrh   w9, [x8, #0x18]
    0x1027f4674 <+8>:  lsr    w8, w9, #3
    0x1027f4678 <+12>: and    x10, x8, #0x1ffc
Target 0: (zig) stopped.
```

**Comment by davidgm94** [comment]
Building an empty Zig file with this build.zig, it gives no error other than a warning from the linker saying it can't find _start.

```
const std = @import("std");

pub fn build(b: *std.build.Builder) void {
    // Standard target options allows the person running `zig build` to choose
    // what target to build for. Here we do not override the defaults, which
    // means any target is allowed, and the default is native. Other options
    // for restricting supported target set are available.
    var disabled_features = std.Target.Cpu.Feature.Set.empty;
    var enabled_features = std.Target.Cpu.Feature.Set.empty;

            const features = std.Target.x86.Feature;
            disabled_features.addFeature(@enumToInt(features.mmx));
            disabled_features.addFeature(@enumToInt(features.sse));
            disabled_features.addFeature(@enumToInt(features.sse2));
            disabled_features.addFeature(@enumToInt(features.avx));
            disabled_features.addFeature(@enumToInt(features.avx2));

            enabled_features.addFeature(@enumToInt(features.soft_float));

    // Standard release options allow the person running `zig build` to select
    // between Debug, ReleaseSafe, ReleaseFast, and ReleaseSmall.
    const mode = b.standardReleaseOptions();

    const exe = b.addExecutable("fault", "src/main.zig");
    exe.setTarget(.{
        .cpu_arch = .x86_64,
        .os_tag = std.Target.Os.Tag.freestanding,
        .abi = std.Target.Abi.none,
        .cpu_features_sub = disabled_features,
        .cpu_features_add = enabled_features,
    });

    exe.setBuildMode(mode);
    exe.install();

    const run_cmd = exe.run();
    run_cmd.step.dependOn(b.getInstallStep());
    if (b.args) |args| {
        run_cmd.addArgs(args);
    }

    const run_step = b.step("run", "Run the app");
    run_step.dependOn(&run_cmd.step);
}
```

**Comment by leecannon** [comment]
@davidgm94 using your build.zig with an empty file with `0.10.0-dev.870+65aa33319` gives:

```
zig build
fault...The following command terminated unexpectedly:
/home/lee/src/github.com/leecannon/zig/build/zig build-exe /home/lee/src/fff/src/main.zig --cache-dir /home/lee/src/fff/zig-cache --global-cache-dir /home/lee/.cache/zig --name fault -target x86_64-freestanding-none -mcpu=x86_64-mmx+soft_float-sse-sse2 --enable-cache
error: the following build command failed with exit code 11:
/home/lee/src/fff/zig-cache/o/202b71519dedb1eed7d7b24918ac680b/build /home/lee/src/github.com/leecannon/zig/build/zig /home/lee/src/fff /home/lee/src/fff/zig-cache /home/lee/.cache/zig
```

running the build command directly shows it is a seg fault:
```
/home/lee/src/github.com/leecannon/zig/build/zig build-exe /home/lee/src/fff/src/main.zig --cache-dir /home/lee/src/fff/zig-cache --global-cache-dir /home/lee/.cache/zig --name fault -target x86_64-freestanding-none -mcpu=x86_64-mmx+soft_float-sse-sse2 --enable-cache
fish: Job 1, '/home/lee/src/github.com/leecan…' terminated by signal SIGSEGV (Address boundary error)
```

**Comment by davidgm94** [comment]
Yeah my bad, I used 0.9.1 zig. I deleted the comment. I apologize.

Still it crashes in another function of LLVM, which is weird.

**Comment by ominitay** [comment]
Apologies, I accidentally left out the target triple. Though it's important to see that this isn't an issue limited to freestanding targets.

**Comment by davidgm94** [comment]
I have been thinking about it and the example that you mentioned shouldn't crash in LLVM (I don't know if I am wrong, but that's my guess). If you remove +soft_float, it doesn't get to LLVM and it gives a compilation error: 
```
./zig-linux-x86_64-0.10.0-dev.871+bb05a8a08/lib/std/start.zig:546:45: error: root source file has no member called 'main'
    switch (@typeInfo(@typeInfo(@TypeOf(root.main)).Fn.return_type.?)) {
```
Shouldn't it give it with +soft_float too? My guess this is another different kind of error.

In fact, if you compile this with 0.9.1, it also seg faults, which indicates that something was wrong there before the bug of this issue was introduced (My guess again is that Zig should give a compiler error before getting to LLVM, but I might be wrong).

Thanks everybody for contributing and I hope my comment helps you too!

**Comment by ominitay** [comment]
> In fact, if you compile this with 0.9.1, it also seg faults, which indicates that something was wrong there before the bug of this issue was introduced (My guess again is that Zig should give a compiler error before getting to LLVM, but I might be wrong).

This also happens with 0.9.0 with the minimum test case, but not with zigvale-barebones, when replacing `kernel.zig` with an empty file. Perhaps that is a separate issue we're running into? I don't know, and I don't know enough to begin to diagnose this, unfortunately. Hopefully somebody more knowledgeable than me has the time for this.

**Comment by ominitay** [comment]
Considering this, this command is the minimal reproduction case for this specific segfault:
```
$ touch empty.zig
$ zig build-exe empty.zig -target x86_64-freestanding-none -mcpu=x86_64+soft_float-sse-sse2
segmentation fault (core dumped)  zig build-exe empty.zig -target x86_64-freestanding-none -mcpu=x86_64+soft_float-sse-sse2
```
This doesn't segfault on non-master Zig versions.

**Comment by davidgm94** [comment]
Alright, thanks. I can take care of this once I figure out how to use llvm-reduce.

**Comment by N00byEdge** [comment]
Still hard blocked by this. Should there be an LLVM issue opened for this? Is zigs fault for sending bad data to LLVM?

**Comment by matu3ba** [comment]
> Still hard blocked by this. Should there be an LLVM issue opened for this?

Next steps
1. Add each function of the commit one by including linkage one to see which crashes with the minimal reproducible example
2. debug stage1
  - trace the panic handlers (likely panics handler is panicing)
  - bisect origin

**Comment by leecannon** [comment]
I've tried to locate the part of commit 9bbd3ab257137c97f695d187436e14c622f877c8 that causes this segfault, using the empty file reproduction provided above.

TLDR: Branching on the expression on [line 158 of `cmp_f80`](https://github.com/ziglang/zig/blob/9bbd3ab257137c97f695d187436e14c622f877c8/lib/std/special/compiler_rt/compareXf2.zig#L158-L160) causes the segfault.

Replacing `cmp_f80` with a dummy, prevents the segfault:
```zig
pub inline fn cmp_f80(comptime RT: type, a: f80, b: f80) RT {
    _ = a;
    _ = b;

    return .Greater;
}
```

After bisecting the function Ive discovered that the below compiles without segfault:

```zig
pub inline fn cmp_f80(comptime RT: type, a: f80, b: f80) RT {
    const a_rep = @ptrCast(*const std.math.F80Repr, &a).*;
    const b_rep = @ptrCast(*const std.math.F80Repr, &b).*;
    const sig_bits = std.math.floatMantissaBits(f80);
    const int_bit = 0x8000000000000000;
    const sign_bit = 0x8000;
    const special_exp = 0x7FFF;

    // START OF CHANGES

    _ = int_bit;
    // If either a or b is NaN, they are unordered.
    // if ((a_rep.exp & special_exp == special_exp and a_rep.fraction ^ int_bit != 0) or
    //     (b_rep.exp & special_exp == special_exp and b_rep.fraction ^ int_bit != 0))
    //     return RT.Unordered;
    
    // END OF CHANGES

    // If a and b are both zeros, they are equal.
    if ((a_rep.fraction | b_rep.fraction) | ((a_rep.exp | b_rep.exp) & special_exp) == 0)
        return .Equal;

    if (@boolToInt(a_rep.exp == b_rep.exp) & @boolToInt(a_rep.fraction == b_rep.fraction) != 0) {
        return .Equal;
    } else if (a_rep.exp & sign_bit != b_rep.exp & sign_bit) {
        // signs are different
        if (@bitCast(i16, a_rep.exp) < @bitCast(i16, b_rep.exp)) {
            return .Less;
        } else {
            return .Greater;
        }
    } else {
        const a_fraction = a_rep.fraction | (@as(u80, a_rep.exp) << sig_bits);
        const b_fraction = b_rep.fraction | (@as(u80, b_rep.exp) << sig_bits);
        if (a_fraction < b_fraction) {
            return .Less;
        } else {
            return .Greater;
        }
    }
}
```

Looking closer at the expression in question, performing the operation does not cause a segfault (first code block below) but using it as the condition for an if statement does segfault (second code block below):
```zig
// No segfault
pub inline fn cmp_f80(comptime RT: type, a: f80, b: f80) RT {
    const a_rep = @ptrCast(*const std.math.F80Repr, &a).*;
    const b_rep = @ptrCast(*const std.math.F80Repr, &b).*;
    const int_bit = 0x8000000000000000;
    const special_exp = 0x7FFF;

    var a1 = (a_rep.exp & special_exp == special_exp and a_rep.fraction ^ int_bit != 0) or
        (b_rep.exp & special_exp == special_exp and b_rep.fraction ^ int_bit != 0);
    std.math.doNotOptimizeAway(a1);

    return .Greater;
}
``` 
```zig
// Segfault
pub inline fn cmp_f80(comptime RT: type, a: f80, b: f80) RT {
    const a_rep = @ptrCast(*const std.math.F80Repr, &a).*;
    const b_rep = @ptrCast(*const std.math.F80Repr, &b).*;
    const int_bit = 0x8000000000000000;
    const special_exp = 0x7FFF;

    var a1 = (a_rep.exp & special_exp == special_exp and a_rep.fraction ^ int_bit != 0) or
        (b_rep.exp & special_exp == special_exp and b_rep.fraction ^ int_bit != 0);
    std.math.doNotOptimizeAway(a1);

    if (a1)
        return RT.Unordered;

    return .Greater;
}
```

**Comment by leecannon** [comment]
I have no idea why that segfaults nor how to move forward, and I definitely don't know enough about f80 and floats in general to write the condition in a different way to see if that would work.
As you are the author of the code any ideas @Vexu?

**Comment by Vexu** [comment]
Looks like you've been debugging an old version, no pointer casts are used as of #10861.

**Comment by ominitay** [comment]
My thought is that they've been debugging with their branch checked out specifically at the problematic commit.

**Comment by leecannon** [comment]
Seems like your right Vexu, in latest aster even commenting the `// If either a or b is NaN, they are unordered.` code out still segfaults. :(

**Comment by N00byEdge** [comment]
@leecannon does the code generated (asm compared with the branch/without it) provide any hints?

**Comment by leecannon** [comment]
@N00byEdge the problem with that is this is a change in compilerrt meaning that there is no change to the zig (stage 1) binary and as having the branch present causes a segfault at compile time there is no binary output to view the assembly of. 

Also as vexu pointed out this code has changed since the original commit that started this problem, due to this commenting out the branch in latest master does not prevent the segfault.

**Comment by leecannon** [comment]
Unfortunately with latest master (0.10.0-dev.2148+67c4b16d6) x86_64 freestanding is still dead in the water due to soft_float not being available. In both stage1 and stage2 (LLVM).

```
$ echo "pub fn main() void {}" > simple.zig
$ # stage 1
$ zig build-exe simple.zig -mcpu=x86_64+soft_float
LLVM Emit Object... [1]    9148 segmentation fault  zig build-exe simple.zig -mcpu=x86_64+soft_float
$ # stage 2 - LLVM
$ zig build-exe simple.zig -mcpu=x86_64+soft_float -fno-stage1 -fLLVM
[1]    10249 segmentation fault  zig build-exe simple.zig -mcpu=x86_64+soft_float -fno-stage1 -fLLVM
$ # stage 2 - no LLVM works
$ zig build-exe simple.zig -mcpu=x86_64+soft_float -fno-stage1 -fno-LLVM
$ ./simple
```

**Comment by N00byEdge** [comment]
Thanks for the continued efforts!

**Comment by leecannon** [comment]
@ominitay i think the title of this issue should be changed as this issue also affects stage2 (llvm backend)

It also essentially makes x86_64 freestanding a dead target, ateast without having the entry point be in assembly to enable SSE (which is not acceptable for "standard" freestanding targets like kernels anyway) before any zig code runs.

**Comment by N00byEdge** [comment]
Yeah, this probably isn't getting enough people knowing about it, since it's literally a tier 1 target. Sadly, I still havn't seen anyone knowing why this is even happening to begin with. Do we still think it's UB somewhere?

**Comment by andrewrk** [comment]
0.10.0 will be shipping with stage2 and will not even have stage1 behind a flag. I have my eye on this issue and will be giving it proper attention once it's stage2 QA time. It's not time yet though. Couple more weeks.

**Comment by N00byEdge** [comment]
Okay, sounds good!

**Comment by ominitay** [comment]
Possibly wants to be tagged as a stage2 bug as well as stage1 then? Though as Andy has an eye on it, it probably doesn't matter.

**Comment by davidgm94** [comment]
Related: https://github.com/ziglang/zig/pull/11844

**Comment by ominitay** [comment]
Doesn't look related: that's not testing anything with freestanding.

**Comment by ominitay** [comment]
Unless this issue has a far wider scope....

**Comment by N00byEdge** [comment]
it's a miscompilation of compiler-rt f80, so might not be limited to soft-float, that might just be what makes it crash.

**Comment by matu3ba** [comment]
Related to #11844, the full list of disabled behavior tests (on llvm windows) is
- thread local variable
- @intToFloat(f80)
- 128-bit multiplication
- remainder division
- float remainder division using @rem
- float modulo division using @mod
- @floor f80
- @ceil f80
- @trunc f80
- @round f80
- fabs
- @mulAdd f80
- vector f80
- saturating multiplication (saturating_arithmetic.zig)
- saturating multiplication (vector.zig) 

I suspect first fixing the "Microsoft X64 Calling Convention" to enable compiler-rt for Windows could help to fix this by at least ruling out problems + few people have a freestanding x86_64 machine lying around for testing.

See also https://en.wikipedia.org/wiki/X86_calling_conventions

**Comment by davidgm94** [comment]
I have reported the issue to LLVM: https://github.com/llvm/llvm-project/issues/56351
I have never reported an issue before to them, so I hope to have done it decently enough. Hopefully it gets fixed soon :)

**Comment by davidgm94** [comment]
Compiling llc with latest master in debug mode triggers an assert instead. I am far from an expert, but based on this assert and the comment on the LLVM issue, it sounds like this code should not be emitted for the desired target:

```
Should not custom lower this!
UNREACHABLE executed at C:\Users\David\dev\llvm-project\llvm\lib\Target\X86\X86ISelLowering.cpp:32029!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: .\\llc.exe reduced-llc.ll
1.      Running pass 'Function Pass Manager' on module 'reduced-llc.ll'.
2.      Running pass 'X86 DAG->DAG Instruction Selection' on function '@compiler_rt.addf3.addf3.64'
Exception Code: 0x80000003
 #0 0x00007ff63d9daa9c HandleAbort C:\Users\David\dev\llvm-project\llvm\lib\Support\Windows\Signals.inc:414:0
 #1 0x00007ffbe436bc31 (C:\Windows\SYSTEM32\ucrtbased.dll+0x6bc31)
 #2 0x00007ffbe436d889 (C:\Windows\SYSTEM32\ucrtbased.dll+0x6d889)
 #3 0x00007ff63d86234f llvm::llvm_unreachable_internal(char const *, char const *, unsigned int) C:\Users\David\dev\llvm-project\llvm\lib\Support\ErrorHandling.cpp:218:0
 #4 0x00007ff63ad9f9c5 llvm::X86TargetLowering::LowerOperation(class llvm::SDValue, class llvm::SelectionDAG &) const C:\Users\David\dev\llvm-project\llvm\lib\Target\X86\X86ISelLowering.cpp:32030:0
 #5 0x00007ff63d62c868 llvm::TargetLowering::LowerOperationWrapper(class llvm::SDNode *, class llvm::SmallVectorImpl<class llvm::SDValue> &, class llvm::SelectionDAG &) const C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGBuilder.cpp:10092:0
 #6 0x00007ff63d7100f0 llvm::DAGTypeLegalizer::CustomLowerNode(class llvm::SDNode *, struct llvm::EVT, bool) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\LegalizeTypes.cpp:933:0
 #7 0x00007ff63d814977 llvm::DAGTypeLegalizer::SplitVectorOperand(class llvm::SDNode *, unsigned int) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\LegalizeVectorTypes.cpp:2676:0
 #8 0x00007ff63d70de7e llvm::DAGTypeLegalizer::run(void) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\LegalizeTypes.cpp:346:0
 #9 0x00007ff63d70d579 llvm::SelectionDAG::LegalizeTypes(void) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\LegalizeTypes.cpp:1060:0
#10 0x00007ff63d3b8bfd llvm::SelectionDAGISel::CodeGenAndEmitDAG(void) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGISel.cpp:838:0
#11 0x00007ff63d3b6237 llvm::SelectionDAGISel::SelectBasicBlock(class llvm::ilist_iterator<struct llvm::ilist_detail::node_options<class llvm::Instruction, 1, 0, void>, 0, 1>, class llvm::ilist_iterator<struct llvm::ilist_detail::node_options<class llvm::Instruction, 1, 0, void>, 0, 1>, bool &) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGISel.cpp:733:0
#12 0x00007ff63d3b5e26 llvm::SelectionDAGISel::SelectAllBasicBlocks(class llvm::Function const &) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGISel.cpp:1639:0
#13 0x00007ff63d3ab8f5 llvm::SelectionDAGISel::runOnMachineFunction(class llvm::MachineFunction &) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGISel.cpp:513:0
#14 0x00007ff63abd31f2 `anonymous namespace'::X86DAGToDAGISel::runOnMachineFunction C:\Users\David\dev\llvm-project\llvm\lib\Target\X86\X86ISelDAGToDAG.cpp:191:0
#15 0x00007ff63c0ec319 llvm::MachineFunctionPass::runOnFunction(class llvm::Function &) C:\Users\David\dev\llvm-project\llvm\lib\CodeGen\MachineFunctionPass.cpp:73:0
#16 0x00007ff63c85443d llvm::FPPassManager::runOnFunction(class llvm::Function &) C:\Users\David\dev\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1430:0
#17 0x00007ff63c8547c7 llvm::FPPassManager::runOnModule(class llvm::Module &) C:\Users\David\dev\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1476:0
#18 0x00007ff63c855d3b `anonymous namespace'::MPPassManager::runOnModule C:\Users\David\dev\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1545:0
#19 0x00007ff63c856b95 llvm::legacy::PassManagerImpl::run(class llvm::Module &) C:\Users\David\dev\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:535:0
#20 0x00007ff63c84e7a2 llvm::legacy::PassManager::run(class llvm::Module &) C:\Users\David\dev\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1673:0
#21 0x00007ff638a92365 compileModule C:\Users\David\dev\llvm-project\llvm\tools\llc\llc.cpp:735:0
#22 0x00007ff638a9388c main C:\Users\David\dev\llvm-project\llvm\tools\llc\llc.cpp:417:0
#23 0x00007ff63ef19159 invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:79:0
#24 0x00007ff63ef1903e __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288:0
#25 0x00007ff63ef18efe __scrt_common_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:331:0
#26 0x00007ff63ef191ee mainCRTStartup D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:17:0
#27 0x00007ffc25b154e0 (C:\Windows\System32\KERNEL32.DLL+0x154e0)
#28 0x00007ffc2680485b (C:\Windows\SYSTEM32\ntdll.dll+0x485b)
```

Additionally, I leave here the information posted on the LLVM tracker.

The stack trace for release mode:

```
fatal error: error in backend: Do not know how to scalarize this operator's operand!

PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: clang -c reduced.ll
1.	Code generation
2.	Running pass 'Function Pass Manager' on module 'reduced.ll'.
3.	Running pass 'X86 DAG->DAG Instruction Selection' on function '@compiler_rt.addf3.addf3.64'
 #0 0x00007faacf797fda (/usr/lib/libLLVM-13.so+0xb97fda)
 #1 0x00007faacf79565f llvm::sys::RunSignalHandlers() (/usr/lib/libLLVM-13.so+0xb9565f)
 #2 0x00007faacf69931f llvm::CrashRecoveryContext::HandleExit(int) (/usr/lib/libLLVM-13.so+0xa9931f)
 #3 0x00007faacf78f5e4 llvm::sys::Process::Exit(int, bool) (/usr/lib/libLLVM-13.so+0xb8f5e4)
 #4 0x000056315d8e2743 (/usr/bin/clang-13+0x10743)
 #5 0x00007faacf6af795 llvm::report_fatal_error(llvm::Twine const&, bool) (/usr/lib/libLLVM-13.so+0xaaf795)
 #6 0x00007faacf6af960 llvm::install_bad_alloc_error_handler(void (*)(void*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool), void*) (/usr/lib/libLLVM-13.so+0xaaf960)
 #7 0x00007faad003d092 (/usr/lib/libLLVM-13.so+0x143d092)
 #8 0x00007faacfff61a6 (/usr/lib/libLLVM-13.so+0x13f61a6)
 #9 0x00007faacfff67d3 llvm::SelectionDAG::LegalizeTypes() (/usr/lib/libLLVM-13.so+0x13f67d3)
#10 0x00007faad0116299 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/usr/lib/libLLVM-13.so+0x1516299)
#11 0x00007faad0119420 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/usr/lib/libLLVM-13.so+0x1519420)
#12 0x00007faad011bc4a (/usr/lib/libLLVM-13.so+0x151bc4a)
#13 0x00007faad289d69f (/usr/lib/libLLVM-13.so+0x3c9d69f)
#14 0x00007faacfbc63be llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/usr/lib/libLLVM-13.so+0xfc63be)
#15 0x00007faacf8fac40 llvm::FPPassManager::runOnFunction(llvm::Function&) (/usr/lib/libLLVM-13.so+0xcfac40)
#16 0x00007faacf8fad84 llvm::FPPassManager::runOnModule(llvm::Module&) (/usr/lib/libLLVM-13.so+0xcfad84)
#17 0x00007faacf8fc61b llvm::legacy::PassManagerImpl::run(llvm::Module&) (/usr/lib/libLLVM-13.so+0xcfc61b)
#18 0x00007faad7c441ed (/usr/lib/libclang-cpp.so.13+0x26441ed)
#19 0x00007faad6cbb4af clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/usr/lib/libclang-cpp.so.13+0x16bb4af)
#20 0x00007faad6f8eaad clang::CodeGenAction::ExecuteAction() (/usr/lib/libclang-cpp.so.13+0x198eaad)
#21 0x00007faad7577f79 clang::FrontendAction::Execute() (/usr/lib/libclang-cpp.so.13+0x1f77f79)
#22 0x00007faad751e2cf clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/libclang-cpp.so.13+0x1f1e2cf)
#23 0x00007faad75c35f0 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib/libclang-cpp.so.13+0x1fc35f0)
#24 0x000056315d8e884c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/bin/clang-13+0x1684c)
#25 0x000056315d8eac2d (/usr/bin/clang-13+0x18c2d)
#26 0x00007faad7287a99 (/usr/lib/libclang-cpp.so.13+0x1c87a99)
#27 0x00007faacf6991b7 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/usr/lib/libLLVM-13.so+0xa991b7)
#28 0x00007faad728e030 (/usr/lib/libclang-cpp.so.13+0x1c8e030)
#29 0x00007faad72a6694 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/usr/lib/libclang-cpp.so.13+0x1ca6694)
#30 0x00007faad72a6ac9 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/usr/lib/libclang-cpp.so.13+0x1ca6ac9)
#31 0x00007faad72a9a42 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/usr/lib/libclang-cpp.so.13+0x1ca9a42)
#32 0x000056315d8e061b main (/usr/bin/clang-13+0xe61b)
#33 0x00007faace429290 (/usr/lib/libc.so.6+0x29290)
#34 0x00007faace42934a __libc_start_main (/usr/lib/libc.so.6+0x2934a)
#35 0x000056315d8e1485 _start (/usr/bin/clang-13+0xf485)
clang-13: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 13.0.1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang-13: note: diagnostic msg: Error generating preprocessed source(s) - no preprocessable inputs.
```

And the reduced IR:

```
; ModuleID = 'llvm.ir'
source_filename = "compiler_rt"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-unknown-eabi"

define internal fastcc fp128 @compiler_rt.addf3.addf3.64() unnamed_addr #0 {
Entry:
  %aAbs = alloca i128, align 16
  br label %BoolOrTrue

BoolOrTrue:                                       ; preds = %Entry
  br label %Then

Then:                                             ; preds = %BoolOrTrue
  br label %Else

Else:                                             ; preds = %Then
  br label %EndIf

EndIf:                                            ; preds = %Else
  br label %Else3

Else3:                                            ; preds = %EndIf
  br label %EndIf4

EndIf4:                                           ; preds = %Else3
  %0 = load i128, i128* %aAbs, align 16
  %1 = load i128, i128* undef, align 16
  %2 = icmp eq i128 %0, %1
  br i1 %2, label %Then5, label %Else8

Then5:                                            ; preds = %EndIf4
  ret fp128 undef

Else8:                                            ; preds = %EndIf4
  ret fp128 undef
}

; Function Attrs: nofree nosync nounwind readnone speculatable willreturn
declare i128 @llvm.ctlz.i128(i128, i1 immarg) #1

attributes #0 = { "target-features"="-16bit-mode,-32bit-mode,-3dnow,-3dnowa,+64bit,-adx,-aes,-amx-bf16,-amx-int8,-amx-tile,-avx,-avx2,-avx512bf16,-avx512bitalg,-avx512bw,-avx512cd,-avx512dq,-avx512er,-avx512f,-avx512ifma,-avx512pf,-avx512vbmi,-avx512vbmi2,-avx512vl,-avx512vnni,-avx512vp2intersect,-avx512vpopcntdq,-avxvnni,-bmi,-bmi2,-branchfusion,-cldemote,-clflushopt,-clwb,-clzero,+cmov,-cx16,+cx8,-enqcmd,-ermsb,-f16c,-false-deps-lzcnt-tzcnt,-false-deps-popcnt,-fast-11bytenop,-fast-15bytenop,-fast-7bytenop,-fast-bextr,-fast-gather,-fast-hops,-fast-lzcnt,-fast-movbe,-fast-scalar-fsqrt,-fast-scalar-shift-masks,-fast-shld-rotate,-fast-variable-crosslane-shuffle,-fast-variable-perlane-shuffle,-fast-vector-fsqrt,-fast-vector-shift-masks,-fma,-fma4,-fsgsbase,-fsrm,+fxsr,-gfni,-hreset,-idivl-to-divb,+idivq-to-divl,-invpcid,-kl,-lea-sp,-lea-uses-ag,-lvi-cfi,-lvi-load-hardening,-lwp,-lzcnt,+macrofusion,+mmx,-movbe,-movdir64b,-movdiri,-mwaitx,+nopl,-pad-short-functions,-pclmul,-pconfig,-pku,-popcnt,-prefer-128-bit,-prefer-256-bit,-prefer-mask-registers,-prefetchwt1,-prfchw,-ptwrite,-rdpid,-rdrnd,-rdseed,-retpoline,-retpoline-external-thunk,-retpoline-indirect-branches,-retpoline-indirect-calls,-rtm,-sahf,-serialize,-seses,-sgx,-sha,-shstk,+slow-3ops-lea,+slow-incdec,-slow-lea,-slow-pmaddwd,-slow-pmulld,-slow-shld,-slow-two-mem-ops,-slow-unaligned-mem-16,-slow-unaligned-mem-32,+soft-float,+sse,+sse2,-sse3,-sse4.1,-sse4.2,-sse4a,-sse-unaligned-mem,-ssse3,-tbm,-tsxldtrk,-uintr,-use-aa,-use-glm-div-sqrt-costs,-vaes,-vpclmulqdq,+vzeroupper,-waitpkg,-wbnoinvd,-widekl,+x87,-xop,-xsave,-xsavec,-xsaveopt,-xsaves" }
attributes #1 = { nofree nosync nounwind readnone speculatable willreturn }
```

**Comment by andrewrk** [comment]
Not planning to fix this in stage1. However the Zig team is here to help you upgrade your projects to stage3 whenever you decide it's time to do the upgrade. I think we are just about coming upon the time where it makes sense to do this.

**Comment by N00byEdge** [comment]
@andrewrk Shouldn't we have #12090 be merged to get stage1 usable until we can build our kernels with stage3 then?

**Comment by ominitay** [comment]
I think we're fine just working with our own builds of that fork tbh. I don't see a reason for that to be even temporarily merged.

**Comment by andrewrk** [comment]
From my perspective we need to get everyone upgraded to stage3 as soon as possible, and any time spent on stage1 is wasted.
